src/commands/Register.ts

```diff
@@ -18,130 +18,215 @@ export type RegisterOptions = {
-          gasLimit: 1e6,
-          gasPrice: options.gasPrice,
-        });
-
-      transactions.push(stakeTx.hash);
-    }
+  const relayAddress = response.relayManagerAddress;
+  const { stake, unstakeDelay, owner } = await relayHub.getStakeInfo(
+    relayAddress
+  );
 
-    if (isSameAddress(owner, from)) {
-      log.info('Relayer already authorized');
-    }
+  log.info('Current stake info:');
+  log.info('Relayer owner: ', owner);
+  log.info('Current unstake delay: ', unstakeDelay.toString());
+  log.info('current stake=', utils.formatUnits(stake.toString(), 'ether'));
 
-    const bal = await this._provider.getBalance(relayAddress);
+  const from = await options.signer.getAddress();
+  if (owner !== constants.AddressZero && !isSameAddress(owner, from)) {
+    throw new Error(`Already owned by ${owner}, our account=${from}`);
+  }
 
-    if (bal.gt(options.funds)) {
-      log.info('Relayer already funded');
-    } else {
-      log.info('Funding relayer');
+  if (unstakeDelay.gte(options.unstakeDelay) && stake.gte(options.stake)) {
+    log.info('Relayer already staked');
+  } else {
+    const stakeValue = options.stake.sub(stake);
+    log.info(
+      `Staking relayer ${utils.formatUnits(stakeValue, 'ether')} RBTC`,
+      stake.eq(constants.Zero)
+        ? ''
+        : ` (already has ${utils.formatUnits(stake, 'ether')} RBTC)`
+    );
 
-      const fundTx = await options.signer.sendTransaction({
-        to: relayAddress,
-        value: options.funds,
+    const stakeTx = await relayHub
+      .connect(options.signer)
+      .stakeForAddress(relayAddress, options.unstakeDelay.toString(), {
+        value: stakeValue,
         gasLimit: 1e6,
         gasPrice: options.gasPrice,
       });
 
-      if (fundTx.hash == null) {
-        throw new Error(`Fund transaction reverted: ${JSON.stringify(fundTx)}`);
-      }
-      transactions.push(fundTx.hash);
-    }
+    transactions.push(stakeTx.hash);
+  }
 
-    await this.waitForRelay(options.relayUrl);
-    log.info('Executed Transactions', transactions);
+  if (isSameAddress(owner, from)) {
+    log.info('Relayer already authorized');
   }
-}
 
-export async function executeRegister(registerOptions?: RegisterOptions) {
+  const bal = await provider.getBalance(relayAddress);
+
+  if (bal.gt(options.funds)) {
+    log.info('Relayer already funded');
+  } else {
+    log.info('Funding relayer');
+
+    const fundTx = await options.signer.sendTransaction({
+      to: relayAddress,
+      value: options.funds,
+      gasLimit: 1e6,
+      gasPrice: options.gasPrice,
+    });
+
+    if (fundTx.hash == null) {
+      throw new Error(`Fund transaction reverted: ${JSON.stringify(fundTx)}`);
+    }
+    transactions.push(fundTx.hash);
+  }
+
+  await waitForRelay(httpClient, options.relayUrl);
+  log.info('Executed Transactions', transactions);
+};
+
+export async function executeRegister() {
+  // FIXME: add registerOptions?: RegisterOptions from either config or cli (which one?)
```

Comment:
should we allow additional config options with the possibility of overriding via env vars for these, or should we follow the original implementation and parse arguments @antomor ?